timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -0,0 +1,740 @@
+    const int N,
+    const int num_target_mols,
+    const int mol_size,
+    const int *__restrict__ mol_atoms_idxs,
+    const int *__restrict__ mol_offsets,
+    const double *__restrict__ per_atom_energies,
+    const double inv_kT,
+    double *__restrict__ log_weights);
+
+template <typename RealType, int THREADS_PER_BLOCK>
+void __global__ k_set_sampled_weight(
+    const int N,
+    const int mol_size,
+    const int num_samples,
+    const int *__restrict__ samples, // [num_samples]
+    const int *__restrict__ target_atoms,
+    const int *__restrict__ mol_offsets,
+    const RealType *__restrict__ per_atom_energies,
+    const RealType inv_kT, // 1 / kT
+    RealType *__restrict__ log_weights) {
+    static_assert(THREADS_PER_BLOCK <= 512 && (THREADS_PER_BLOCK & (THREADS_PER_BLOCK - 1)) == 0);
+    assert(num_samples == 1);
+    assert(gridDim.x == 1);
+    __shared__ __int128 accumulators[THREADS_PER_BLOCK];
+
+    int sample_idx = samples[0];
+    int min_atom_idx = target_atoms[0];
+    int max_atom_idx = target_atoms[mol_size - 1];
+
+    int atom_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    __int128 accumulator = 0;
+    // Zero all of the accumulators
+    while (atom_idx < N) {
+        if (atom_idx < min_atom_idx || atom_idx > max_atom_idx) {
+            for (int i = 0; i < mol_size; i++) {
+                accumulator += FLOAT_TO_FIXED_ENERGY<RealType>(inv_kT * per_atom_energies[i * N + atom_idx]);
+            }
+        }
+        atom_idx += gridDim.x * blockDim.x;
+    }
+    accumulators[threadIdx.x] = accumulator;
+    __syncthreads();
+    block_energy_reduce<THREADS_PER_BLOCK>(accumulators, threadIdx.x);
+    if (threadIdx.x == 0) {
+        log_weights[sample_idx] =
+            fixed_point_overflow(accumulators[0]) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(accumulators[0]);
+    }
+}
+
+template void __global__ k_set_sampled_weight<float, 512>(
+    const int N,
+    const int mol_size,
+    const int num_samples,
+    const int *__restrict__ samples, // [num_samples]
+    const int *__restrict__ target_atoms,
+    const int *__restrict__ mol_offsets,
+    const float *__restrict__ per_atom_energies,
+    const float inv_kT,
+    float *__restrict__ log_weights);
+template void __global__ k_set_sampled_weight<double, 512>(
+    const int N,
+    const int mol_size,
+    const int num_samples,
+    const int *__restrict__ samples, // [num_samples]
+    const int *__restrict__ target_atoms,
+    const int *__restrict__ mol_offsets,
+    const double *__restrict__ per_atom_energies,
+    const double inv_kT,
+    double *__restrict__ log_weights);
+
+template <typename RealType>
+void __global__ k_compute_centroid_of_atoms(
+    const int num_atoms,
+    const int *__restrict__ atom_idxs, // [num_atoms]
+    const double *__restrict__ coords, // [N, 3]
+    RealType *__restrict__ centroid    // [3]
+) {
+    __shared__ unsigned long long fixed_centroid[3];
```

Comment:
why is centroid computed in floating point? (to avoid order warp ordering dependence within a single block?)

Reply:
Do you mean why is it not computed in floating point? Its computed in long long to allow atomic adds with deterministic results.

Though did find a bug because of this comment, should only allow a single block.

Reply:
edit: yeah

Reply:
(explanation makes sense)